File: server/middleware/auth_middleware.py
"""
Authentication and Authorization Middleware

Provides:
- Token verification
- Role-based access control (RBAC)
- Decorators for protecting routes
"""

from functools import wraps
from flask import request, jsonify
import jwt
import os
from pymongo import MongoClient

# MongoDB Connection
MONGO_URI = os.getenv('MONGO_URI', 'mongodb://localhost:27017')
client = MongoClient(MONGO_URI)
db = client.get_database('builtcred')
users_collection = db['users']

JWT_SECRET = os.getenv('JWT_SECRET', 'your-secret-key-change-in-production')
JWT_ALGORITHM = 'HS256'

def verify_token(token):
    """Verify JWT token and return user data"""
    try:
        payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        user_id = payload.get('user_id')
        
        if not user_id:
            return None, 'Invalid token'
        
        # Get user from database
        from bson.objectid import ObjectId
        user = users_collection.find_one({'_id': ObjectId(user_id)})
        
        if not user:
            return None, 'User not found'
        
        return user, None
    except jwt.ExpiredSignatureError:
        return None, 'Token has expired'
    except jwt.InvalidTokenError:
        return None, 'Invalid token'
    except Exception as e:
        return None, str(e)

def get_token_from_request():
    """Extract JWT token from Authorization header"""
    auth_header = request.headers.get('Authorization', '')
    
    if not auth_header:
        return None
    
    try:
        # Expected format: "Bearer <token>"
        parts = auth_header.split()
        if len(parts) != 2 or parts[0].lower() != 'bearer':
            return None
        return parts[1]
    except Exception:
        return None

def login_required(f):
    """Decorator to require authentication"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = get_token_from_request()
        
        if not token:
            return jsonify({'error': 'Missing authorization token'}), 401
        
        user, error = verify_token(token)
        
        if error:
            return jsonify({'error': error}), 401
        
        if not user.get('is_active', True):
            return jsonify({'error': 'Account is inactive'}), 403
        
        if user.get('status') != 'approved':
            return jsonify({'error': 'Account not approved'}), 403
        
        # Attach user to request for use in the route
        request.current_user = user
        return f(*args, **kwargs)
    
    return decorated_function

def role_required(*required_roles):
    """Decorator to require specific roles"""
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            token = get_token_from_request()
            
            if not token:
                return jsonify({'error': 'Missing authorization token'}), 401
            
            user, error = verify_token(token)
            
            if error:
                return jsonify({'error': error}), 401
            
            if not user.get('is_active', True):
                return jsonify({'error': 'Account is inactive'}), 403
            
            if user.get('status') != 'approved':
                return jsonify({'error': 'Account not approved'}), 403
            
            user_role = user.get('role', 'user')
            
            if user_role not in required_roles:
                return jsonify({
                    'error': f'Insufficient permissions. Required roles: {", ".join(required_roles)}'
                }), 403
            
            # Attach user to request for use in the route
            request.current_user = user
            return f(*args, **kwargs)
        
        return decorated_function
    return decorator

def admin_required(f):
    """Decorator to require admin role"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = get_token_from_request()
        
        if not token:
            return jsonify({'error': 'Missing authorization token'}), 401
        
        user, error = verify_token(token)
        
        if error:
            return jsonify({'error': error}), 401
        
        if not user.get('is_active', True):
            return jsonify({'error': 'Account is inactive'}), 403
        
        if user.get('status') != 'approved':
            return jsonify({'error': 'Account not approved'}), 403
        
        if user.get('role') != 'admin':
            return jsonify({'error': 'Admin access required'}), 403
        
        # Attach user to request for use in the route
        request.current_user = user
        return f(*args, **kwargs)
    
    return decorated_function

File: server/middleware/authMiddleware.js
const jwt = require('jsonwebtoken');
const { getDb } = require('../db');
const { ObjectId } = require('mongodb');

const JWT_SECRET = process.env.JWT_SECRET || 'your-secret-key-change-in-production';

async function verifyToken(token) {
    try {
        const payload = jwt.verify(token, JWT_SECRET);
        const userId = payload.user_id;
        if (!userId) return [null, 'Invalid token'];
        const db = getDb();
        const user = await db.collection('users').findOne({ _id: new ObjectId(userId) });
        if (!user) return [null, 'User not found'];
        return [user, null];
    } catch (err) {
        if (err.name === 'TokenExpiredError') return [null, 'Token has expired'];
        return [null, err.message || 'Invalid token'];
    }
}

function getTokenFromRequest(req) {
    const authHeader = req.headers['authorization'] || '';
    if (!authHeader) return null;
    const parts = authHeader.split(' ');
    if (parts.length !== 2 || parts[0].toLowerCase() !== 'bearer') return null;
    return parts[1];
}

function loginRequired(handler) {
    return async (req, res, next) => {
        try {
            const token = getTokenFromRequest(req);
            if (!token) return res.status(401).json({ error: 'Missing authorization token' });
            const [user, error] = await verifyToken(token);
            if (error) return res.status(401).json({ error });
            if (!user.is_active) return res.status(403).json({ error: 'Account is inactive' });
            if (user.status !== 'approved') return res.status(403).json({ error: 'Account not approved' });
            req.currentUser = user;
            return handler(req, res, next);
        } catch (err) {
            next(err);
        }
    };
}

function roleRequired(...roles) {
    return (handler) => async (req, res, next) => {
        try {
            const token = getTokenFromRequest(req);
            if (!token) return res.status(401).json({ error: 'Missing authorization token' });
            const [user, error] = await verifyToken(token);
            if (error) return res.status(401).json({ error });
            if (!user.is_active) return res.status(403).json({ error: 'Account is inactive' });
            if (user.status !== 'approved') return res.status(403).json({ error: 'Account not approved' });
            if (!roles.includes(user.role)) return res.status(403).json({ error: `Insufficient permissions. Required roles: ${roles.join(', ')}` });
            req.currentUser = user;
            return handler(req, res, next);
        } catch (err) {
            next(err);
        }
    };
}

function advocateRequired(handler) {
    return async (req, res, next) => {
        try {
            const token = getTokenFromRequest(req);
            if (!token) return res.status(401).json({ error: 'Missing authorization token' });
            const [user, error] = await verifyToken(token);
            if (error) return res.status(401).json({ error });
            if (!user.is_active) return res.status(403).json({ error: 'Account is inactive' });
            if (user.status !== 'approved') return res.status(403).json({ error: 'Account not approved' });
            if (!user.advocate_type) return res.status(403).json({ error: 'Advocate access required' });
            req.currentUser = user;
            return handler(req, res, next);
        } catch (err) {
            next(err);
        }
    };
}

function advocateOwnsResource(handler) {
    return async (req, res, next) => {
        try {
            const token = getTokenFromRequest(req);
            if (!token) return res.status(401).json({ error: 'Missing authorization token' });
            const [user, error] = await verifyToken(token);
            if (error) return res.status(401).json({ error });
            if (!user.is_active) return res.status(403).json({ error: 'Account is inactive' });
            if (user.status !== 'approved') return res.status(403).json({ error: 'Account not approved' });

            const resource_user_id = req.params.user_id || req.params.advocate_id || req.query.user_id || req.query.advocate_id || (req.body && (req.body.user_id || req.body.advocate_id));
            if (resource_user_id) {
                const current_user_id = String(user._id);
                if (resource_user_id !== current_user_id) return res.status(403).json({ error: 'You do not have permission to access this resource' });
            }

            req.currentUser = user;
            return handler(req, res, next);
        } catch (err) {
            next(err);
        }
    };
}

module.exports = { verifyToken, getTokenFromRequest, loginRequired, roleRequired, advocateRequired, advocateOwnsResource };
